test/integration/deployment/deployment_test.go

```diff
@@ -498,3 +499,111 @@ func TestRollbackDeploymentRSNoRevision(t *testing.T) {
+	s, closeFn, rm, dc, informers, c := dcSetup(t)
+	defer closeFn()
+	name := "test-overlapping-deployments"
+	ns := framework.CreateTestingNamespace(name, s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+
+	replicas := int32(1)
+	firstDeploymentName := "first-deployment"
+	secondDeploymentName := "second-deployment"
+	testers := []*deploymentTester{
+		{t: t, c: c, deployment: newDeployment(firstDeploymentName, ns.Name, replicas)},
+		{t: t, c: c, deployment: newDeployment(secondDeploymentName, ns.Name, replicas)},
+	}
+
+	// Start informer and controllers
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	go dc.Run(5, stopCh)
+
+	// Create 2 deployments with overlappings selectors
+	var err error
+	for _, tester := range testers {
+		tester.deployment, err = c.ExtensionsV1beta1().Deployments(ns.Name).Create(tester.deployment)
+		dname := tester.deployment.Name
+		if err != nil {
+			t.Fatalf("failed to create deployment %q: %v", dname, err)
+		}
+		// Wait for the deployment to be updated to revision 1
+		if err = tester.waitForDeploymentRevisionAndImage("1", fakeImage); err != nil {
+			t.Fatalf("failed to update deployment %q to revision 1: %v", dname, err)
+		}
+		// Make sure the deployment to complete while manually marking its pods as ready at the same time
+		if err = tester.waitForDeploymentCompleteAndMarkPodsReady(); err != nil {
+			t.Fatalf("deployment %q failed to complete: %v", dname, err)
+		}
+	}
+
+	// Both deployments should proceed independently
+	rsSelector, err := metav1.LabelSelectorAsSelector(testers[0].deployment.Spec.Selector)
+	if err != nil {
+		t.Fatalf("failed to parse deployment selector: %v", err)
+	}
+	rsList, err := c.ExtensionsV1beta1().ReplicaSets(ns.Name).List(metav1.ListOptions{LabelSelector: rsSelector.String()})
+	if err != nil {
+		t.Fatalf("failed to list replicasets of deployments %q and %q in namespace %q: %v", firstDeploymentName, secondDeploymentName, ns.Name, err)
+	}
+	rss := rsList.Items
+	if len(rss) != 2 {
+		t.Fatalf("expected 2 replicasets to be found, but found %d", len(rss))
+	}
+
+	firstController := metav1.GetControllerOf(&rss[0])
+	if firstController == nil {
+		t.Fatalf("replicaset %q is not owned by any deployment", rss[0].Name)
+	}
+	secondController := metav1.GetControllerOf(&rss[1])
+	if secondController == nil {
+		t.Fatalf("replicaset %q is not owned by any deployment", rss[1].Name)
+	}
+
+	firstDeploymentUID := testers[0].deployment.UID
+	secondDeploymentUID := testers[1].deployment.UID
+	if !((firstController.UID == firstDeploymentUID && secondController.UID == secondDeploymentUID) ||
+		(firstController.UID == secondDeploymentUID && secondController.UID == firstDeploymentUID)) {
+		t.Fatalf("expected replicasets %q and %q are owned by deployments %q and %q, but found deployments %q and %q", rss[0].Name, rss[1].Name, firstDeploymentName, secondDeploymentName, firstController.Name, secondController.Name)
```

Comment:
If you create a map of [controllerRef: RS], you don't need this complex logic here. Just range through both deployments and check that they exist in the map. You can easily see which RS each deployment has with this map too.

Reply:
To speed up code review process, i wish to ensure we are on the same page via comments. Do you mean something like this: 
```
firstController := metav1.GetControllerOf(&rss[0])
if firstController == nil {
	...
}
secondController := metav1.GetControllerOf(&rss[1])
if secondController == nil {
	...
}

// Set up [controllerRef's UID: replicaset] mapping for future usage
mapping := map[types.UID]*v1beta1.ReplicaSet{firstController.UID: rss[0], secondController.UID: rss[1]} 

// Verify replicasets are owned by the created deployments
for _, tester := range testers {
	if _, ok := dict[tester.deployment.UID]; !ok {
    		t.Fatalf("expected deployment %q owns replicaset %q or %q, but it does not", tester.deployment.Name, rss[0].Name, rss[1].Name)
	}
}
```